oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/GrantedOAuth2AccessToken.java

```diff
@@ -0,0 +1,407 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+
+/**
+ * Defines a structure of the Access Token Response, as per
+ * <a href="https://tools.ietf.org/html/rfc6749#section-5.1">[RFC6749], Section 5.1</a>.
+ */
+public class GrantedOAuth2AccessToken implements Serializable {
+
+    private static final long serialVersionUID = 8698118404098897958L;
+
+    /**
+     * Creates a new {@link GrantedOAuth2AccessToken} based on the {@code JSON}-formatted raw response body and
+     * optional raw formatted {@code scope} used to request the token.
+     * @param rawResponse {@code JSON}-formatted raw response body.
+     * @param requestScope OPTIONAL. A list of space-delimited, case-sensitive strings.
+     *                     The strings are defined by the authorization server.
+     *                     The authorization server MAY fully or partially ignore the scope requested by the
+     *                     client, based on the authorization server policy or the resource owner's
+     *                     instructions. If the issued access token scope is different from the one requested
+     *                     by the client, the authorization server MUST include the "scope" response
+     *                     parameter to inform the client of the actual scope granted.
+     *                     If the client omits the scope parameter when requesting authorization, the
+     *                     authorization server MUST either process the request using a pre-defined default
+     *                     value or fail the request indicating an invalid scope.
+     * @return A new instance of {@link GrantedOAuth2AccessToken}.
+     */
+    public static GrantedOAuth2AccessToken of(String rawResponse, @Nullable String requestScope) {
+        return GrantedOAuth2AccessTokenBuilder.of(rawResponse, requestScope);
+    }
+
+    /**
+     * Creates a new {@link GrantedOAuth2AccessTokenBuilder} to build a new {@link GrantedOAuth2AccessToken} and
+     * supplied it with a value of {@code access_token} Access Token response field.
+     * @return A new instance of {@link GrantedOAuth2AccessTokenBuilder}.
+     */
+    public static GrantedOAuth2AccessTokenBuilder builder(String accessToken) {
+        return new GrantedOAuth2AccessTokenBuilder(accessToken);
+    }
+
+    public static final String ACCESS_TOKEN = "access_token";
+    public static final String REFRESH_TOKEN = "refresh_token";
+    public static final String SCOPE = "scope";
+    public static final String TOKEN_TYPE = "token_type";
+    public static final String EXPIRES_IN = "expires_in";
```

Comment:
Should these be public? If we want to hide these fields, we could add `private` or remove `public`. If it's accessed by the classes in other package, we could extract these constants into a class like `OAuth2Constants` in `internal.common.auth.oauth2` package to hide them from the public API.

Reply:
Those were not public before, but since you've asked me to put various requests into different packages, I've made those public. Weather or not those constants extracted into a separate class, those would remain public.

Reply:
Yeah, what I'm suggesting is to extract these public constants into a new class like `OAuth2Constants` in an internal package - `internal.common.auth.oauth2`. We exclude the internal packages from our Javadoc and a user usually knows it's not a good idea to use the classes in an 'internal' package, so we are sort of 'hiding' it from the public API, although not perfect.

Reply:
moved to internal as suggested